src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -198,820 +542,1177 @@ private void checkDataLength(int ... lengths) {
-                        // If the remaining in buffer + data does not fill a
-                        // block, complete the ghash operation
-                        int l = ct.remaining();
-                        ct.get(block, over, l);
-                        ghashAllToS.doLastBlock(ByteBuffer.wrap(block), over + l);
-                    }
+            processAAD();
+            if (src.remaining() > 0) {
+                // If there is an array, use that to avoid the extra copy to
+                // take the src data out of the bytebuffer.
+                if (src.hasArray()) {
+                    doUpdate(src.array(), src.arrayOffset() + src.position(),
+                        src.remaining(), null, 0);
+                    src.position(src.limit());
                 } else {
-                    // data is empty, so complete the ghash op with the
-                    // remaining buffer
-                    ghashAllToS.doLastBlock(buffer, buffer.remaining());
+                    byte[] b = new byte[src.remaining()];
+                    src.get(b);
+                    initBuffer(b.length);
+                    try {
+                        ibuffer.write(b);
+                    } catch (IOException e) {
+                        throw new ProviderException(
+                            "Unable to add remaining input to the buffer", e);
+                    }
                 }
             }
-            // Prepare buffer for decryption
-            buffer.flip();
+            return 0;
         }
 
-        if (ct.remaining() > 0) {
-            ghashAllToS.doLastBlock(ct, ct.remaining());
-        }
-        // Prepare buffer for decryption if available
-        ct.reset();
+        /**
+         * Use any data from ibuffer and 'in' to first verify the auth tag. If
+         * the tag is valid, decrypt the data.
+         */
+        @Override
+        public int doFinal(byte[] in, int inOfs, int inLen, byte[] out,
+            int outOfs) throws IllegalBlockSizeException, AEADBadTagException,
+            ShortBufferException {
+            GHASH save = null;
+
+            int bufLen = getBufferedLength();
+            int len = inLen + bufLen;
+            if (len < tagLenBytes) {
+                throw new AEADBadTagException("Input too short - need tag");
+            }
 
-        byte[] block = getLengthBlock(sizeOfAAD, len);
-        ghashAllToS.update(block);
-        block = ghashAllToS.digest();
-        GCTR gctrForSToTag = new GCTR(embeddedCipher, this.preCounterBlock);
-        gctrForSToTag.doFinal(block, 0, tagLenBytes, block, 0);
+            if (len - tagLenBytes > out.length - outOfs) {
+                save = ghashAllToS.clone();
+            }
 
-        // check entire authentication tag for time-consistency
-        int mismatch = 0;
-        for (int i = 0; i < tagLenBytes; i++) {
-            mismatch |= tag.get() ^ block[i];
-        }
+            checkDataLength(len - tagLenBytes);
```

Comment:
checkDataLength is subtracting the length's from the max.  The check at 1422 would fail because the max would be negative and the processed would be 0.  I don't see it always passing